### PR TITLE
crimson/osd: Generalize loading obc for head/clone

### PR DESCRIPTION
**Pull request by Matan-B**
* `load_head_obc` is changed to `load_obc` and can be used to load clone obc's as well.
* `get_or_load_obc` is introduced  and used both by `with_head/clone_obc` avoiding boilerplates.

 (This change will also be useful for `CEPH_OSD_OP_ROLLBACK` implementation).

Signed-off-by: Matan Breizman <mbreizma@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/pg.cc [hunk 1]
```diff
@@ -1146,6 +1129,25 @@ PG::load_obc(ObjectContextRef obc)
...
 
+template<RWState::State State>
+PG::load_obc_iertr::future<crimson::osd::ObjectContextRef>
+PG::get_or_load_obc(
+    crimson::osd::ObjectContextRef obc,
+    bool existed)
+{
+  auto loaded = load_obc_iertr::make_ready_future<ObjectContextRef>(obc);
+  if (existed) {
+    logger().debug("with_clone_obc: found {} in cache", obc->get_oid());
```
Comment (rzarzynski) [inline src/crimson/osd/pg.cc:1140]: Oops, old name buried in the debug.

File: src/crimson/osd/pg.cc [hunk 2]
```diff
@@ -1146,6 +1129,25 @@ PG::load_obc(ObjectContextRef obc)
...
+  } else {
+    logger().debug("with_clone_obc: cache miss on {}", obc->get_oid());
```
Comment (rzarzynski) [inline src/crimson/osd/pg.cc:1142]: Same as above.

**Comment by athanatos** [comment]
This approach basically requires all FuturizedStore implementations to be accessed by the OSD through a seastar::sharded<ShardStores> container.  This already isn't a good fit for alienstore -- which doesn't need it, and probably won't be a good fit for SeaStore -- which will need finer control over shared data structures.  It may make sense for CyanStore, but I'd like you do to all of this within the CyanStore implementation without exposing the seastar::sharded<> container to the osd.

**Comment by Matan-B** [comment]
> Run through crimson-rados suite first?

LGTM, no new failrues.
https://pulpito.ceph.com/matan-2022-10-23_13:16:25-crimson-rados-main-distro-crimson-smithi/

**Review by athanatos — APPROVED** [review]

**Review by xxhdx1985126 — APPROVED** [review]

**Pull request merged**